File: app/src/main/java/com/u1fukui/bbs/viewmodel/ThreadListViewModel.java
package com.u1fukui.bbs.viewmodel;

import android.databinding.ObservableArrayList;
import android.databinding.ObservableBoolean;
import android.databinding.ObservableList;

import com.u1fukui.bbs.App;
import com.u1fukui.bbs.model.BbsThread;
import com.u1fukui.bbs.model.ThreadListResponse;
import com.u1fukui.bbs.repository.ThreadListRepository;
import com.u1fukui.bbs.view.customview.ErrorView;
import com.u1fukui.bbs.view.helper.LoadingManager;
import com.u1fukui.bbs.view.helper.ThreadListNavigator;

import java.util.ArrayList;
import java.util.List;

import io.reactivex.SingleObserver;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.annotations.NonNull;
import io.reactivex.disposables.Disposable;
import io.reactivex.schedulers.Schedulers;
import lombok.Getter;

public class ThreadListViewModel implements ViewModel, ErrorView.ErrorViewListener {

    //region DataBinding
    public final LoadingManager loadingManager = new LoadingManager();

    public final ObservableBoolean refreshing = new ObservableBoolean(false);
    //endregion

    @Getter
    private ObservableList<ThreadViewModel> threadViewModelList = new ObservableArrayList<>();

    private boolean isThreadListCompleted;

    private final ThreadListRepository repository;

    private final ThreadListNavigator navigator;

    public ThreadListViewModel(ThreadListRepository repository, ThreadListNavigator navigator) {
        this.repository = repository;
        this.navigator = navigator;
    }

    //region Databinding
    public void onSwipeRefresh() {
        refreshing.set(true);
        fetchThreadList(0);
    }
    //endregion

    public void start() {
        if (threadViewModelList.isEmpty()) {
            fetchThreadList(0);
        }
    }

    public void loadNextPage() {
        if (threadViewModelList.isEmpty()) {
            return;
        }
        ThreadViewModel thread = threadViewModelList.get(threadViewModelList.size() - 1);
        fetchThreadList(thread.bbsThread.id);
    }

    private void fetchThreadList(final long lastId) {
        if (loadingManager.isLoading() || isThreadListCompleted) {
            refreshing.set(false);
            return;
        }
        loadingManager.startLoading();

        repository.fetchThreadList(lastId)
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(new SingleObserver<ThreadListResponse>() {
                    @Override
                    public void onSubscribe(@NonNull Disposable d) {
                        // nop
                    }

                    @Override
                    public void onSuccess(@NonNull ThreadListResponse response) {
                        List<ThreadViewModel> viewModelList = new ArrayList<>();
                        for (BbsThread thread : response.threadList) {
                            viewModelList.add(new ThreadViewModel(navigator, thread));
                        }

                        isThreadListCompleted = response.isCompleted;
                        if (lastId == 0) {
                            threadViewModelList.clear();
                        }
                        threadViewModelList.addAll(viewModelList);
                        loadingManager.showContentView();
                        refreshing.set(false);
                    }

                    @Override
                    public void onError(@NonNull Throwable e) {
                        if (lastId == 0) {
                            loadingManager.showErrorView(e);
                        } else {
                            App.getInstance().getToastUtils().showToast("エラー"); //TODO: エラーメッセージ
                            isThreadListCompleted = true;
                            loadingManager.finishLoading();
                        }
                        refreshing.set(false);

                    }
                });
    }

    @Override
    public void destroy() {
    }

    @Override
    public void onClickReloadButton() {
        fetchThreadList(0);
    }
}

File: app/src/main/java/com/u1fukui/bbs/ui/main/ThreadListViewModel.kt
package com.u1fukui.bbs.ui.main

import androidx.lifecycle.LifecycleOwner
import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.Observer
import androidx.lifecycle.ViewModel
import androidx.lifecycle.ViewModelProvider
import androidx.paging.LivePagedListBuilder
import androidx.paging.PagedList
import com.u1fukui.bbs.App
import com.u1fukui.bbs.customview.ErrorView
import com.u1fukui.bbs.helper.LoadingManager
import com.u1fukui.bbs.paging.Status
import com.u1fukui.bbs.paging.thread.ThreadDataSourceFactory
import com.u1fukui.bbs.repository.thread_list.ThreadListRepository
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlin.coroutines.CoroutineContext

class ThreadListViewModel(
    private val owner: LifecycleOwner,
    private val repository: ThreadListRepository,
    private val navigator: ThreadListNavigator
) : ViewModel(), ErrorView.ErrorViewListener, CoroutineScope {

    //region CoroutineScope
    override val coroutineContext: CoroutineContext
        get() = Dispatchers.Main + job
    //endregion

    //region DataBinding
    val loadingManager = LoadingManager()
    val refreshing = MutableLiveData<Boolean>()
    //endregion

    //    internal val threadBindingModelList: ObservableList<ThreadBindingModel> = ObservableArrayList()

    internal val threadBindingModelList: LiveData<PagedList<ThreadBindingModel>>

    private val job = Job()

    private val factory = ThreadDataSourceFactory(repository, this)

    init {
        val config = PagedList.Config.Builder()
            .setInitialLoadSizeHint(PAGE_SIZE)
            .setPageSize(PAGE_SIZE)
            .setEnablePlaceholders(false)
            .setPrefetchDistance(5)
            .build()

        threadBindingModelList = LivePagedListBuilder(
            factory.map { ThreadBindingModel(navigator, it) },
            config
        ).build()

        observeNetworkState()
    }

    private fun observeNetworkState() {
        factory.source.initialLoad.observe(owner, Observer {
            when (it?.status) {
                Status.RUNNING -> loadingManager.startLoading()
                Status.SUCCESS -> {
                    loadingManager.showContentView()
                    refreshing.postValue(false)
                }
                Status.FAILED -> {
                    loadingManager.showErrorView(it.throwable!!)
                    refreshing.postValue(false)
                }
            }
        })

        factory.source.networkState.observe(owner, Observer {
            when (it?.status) {
                Status.RUNNING -> loadingManager.startLoading()
                Status.SUCCESS -> loadingManager.finishLoading()
                Status.FAILED -> {
                    loadingManager.finishLoading()
                    App.getToastUtils().showApiError(it.throwable!!)
                }
            }
        })
    }

    //region Databinding
    fun onSwipeRefresh() {
        refreshing.postValue(true)
        factory.source.invalidate()
    }
    //endregion

    override fun onCleared() {
        job.cancel()
        super.onCleared()
    }

    override fun onClickReloadButton() {
        factory.source.invalidate()
    }

    class Factory(
        private val lifecycle: LifecycleOwner,
        private val repository: ThreadListRepository,
        private val navigator: ThreadListNavigator
    ) : ViewModelProvider.NewInstanceFactory() {

        @Suppress("UNCHECKED_CAST")
        override fun <T : ViewModel> create(modelClass: Class<T>): T = ThreadListViewModel(
            lifecycle,
            repository,
            navigator
        ) as T
    }

    companion object {
        private const val PAGE_SIZE = 20
    }
}
